src/iris/api.py

```diff
@@ -5604,6 +5610,303 @@ def on_get(self, req, resp):
+                # delete IMP_bucket_changes request for teh bucket we just reassigned
+                cursor.execute("DELETE FROM IMP_bucket_changes WHERE node_id = %s AND bucket_id = %s",
+                               (row["node_id"], row["bucket_id"]))
+            connection.commit()
+
+            # check if there are are nodes who's cluster membership TTL has expired and redistribute their buckets
+
+            # if a node has gone more than TTL seconds without checking in kick it out of the cluster
+            cursor.execute(
+                "SELECT * FROM IMP_cluster_members WHERE TIMESTAMPDIFF(SECOND,last_modified, NOW()) >= %s", self.sender_ttl)
+            dead_nodes = cursor.fetchall()
+
+            if len(dead_nodes) > 0:
+                cursor.execute(
+                    "SELECT * FROM IMP_cluster_members WHERE TIMESTAMPDIFF(SECOND,last_modified, NOW()) < %s", self.sender_ttl)
+                alive_nodes = cursor.fetchall()
+
+                # apply any changes that were already pending for the dead nodes
+                for dead_node in dead_nodes:
+                    cursor.execute(
+                        "SELECT * FROM IMP_bucket_changes WHERE bucket_id IN (SELECT bucket_id FROM IMP_bucket_assignments WHERE node_id = %s)", dead_node["node_id"])
+                    bucket_changes_results = cursor.fetchall()
+                    for row in bucket_changes_results:
+                        # release buckets to the node that requested them
+                        cursor.execute("UPDATE IMP_bucket_assignments SET node_id = %s WHERE bucket_id = %s", (
+                            row["node_id"], row["bucket_id"]))
+                        # delete IMP_bucket_changes request for teh bucket we just reassigned
+                        cursor.execute("DELETE FROM IMP_bucket_changes WHERE node_id = %s AND bucket_id = %s", (
+                            row["node_id"], row["bucket_id"]))
+
+                # give away the rest of the dead nodes' buckets in round robin fashion
+                cursor.execute("SELECT * FROM IMP_bucket_assignments WHERE IMP_bucket_assignments.node_id IN (SELECT IMP_cluster_members.node_id FROM IMP_cluster_members WHERE TIMESTAMPDIFF(SECOND,last_modified, NOW()) >= %s)", self.sender_ttl)
+                dead_node_buckets = cursor.fetchall()
+
+                i = 0
+                for bucket in dead_node_buckets:
+                    if i >= len(alive_nodes):
+                        i = 0
+                    cursor.execute("UPDATE IMP_bucket_assignments SET node_id = %s WHERE bucket_id = %s",
+                                   (alive_nodes[i]["node_id"], bucket["bucket_id"]))
+                    i += 1
+
+                #  delete dead nodes from cluster membership
+                cursor.execute(
+                    "DELETE FROM IMP_cluster_members WHERE TIMESTAMPDIFF(SECOND,last_modified, NOW()) >= %s", self.sender_ttl)
```

Comment:
Running this delete might get you different result from the previous select because some time has passed. You should just delete based off dead_nodes.